File: 프로그래머스/lv3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;
class Solution {
    public int solution(int[][] jobs) {
        int answer = 0;
        Arrays.sort(jobs,(o1,o2)->o1[0]-o2[0]); //오름차순 정렬 
        PriorityQueue<int[]> pq=new PriorityQueue<>((o1,o2)->o1[1]-o2[1]);
        int idx=0;
        int cnt=0;
        int total=0;
        int end=0;

        while(cnt<jobs.length){
            while(idx<jobs.length && jobs[idx][0]<=end){
                pq.offer(jobs[idx++]);
            }
            if(pq.isEmpty()) end = jobs[idx][0]; //end= 현재작업소요시간 
            else{
                int[] cur = pq.poll();
                total += cur[1]+end-cur[0];
                end+=cur[1];
                cnt++;
            }
        }
        
        return total/jobs.length;
    }
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
from heapq import heappush,heappop
from collections import deque
def solution(jobs):
    answer = 0
    jobs.sort()
    end = 0
    
    q = deque([(t,s) for s,t in jobs])
    hq = []
    heappush(hq,(q.popleft()))
    while hq:
        t,s = heappop(hq)
        end = max(end+t,s+t)
        answer+=end-s
        cnt = len(q)
        while q and q[0][1]<=end:
            heappush(hq,q.popleft())
        if q and not hq:
            heappush(hq,q.popleft())
                
            

    return answer//len(jobs)